File: api/src/main/java/org/apache/gravitino/job/ShellJobTemplate.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.job;

import com.google.common.collect.ImmutableList;
import java.util.List;
import java.util.Objects;

/**
 * Represents a job template for executing shell scripts. This class extends the JobTemplate class
 * and provides functionality specific to shell job templates, including a list of scripts to be
 * executed.
 *
 * <p>Scripts are a list of shell files that will be leveraged by the "executable". Scripts must be
 * put in the place where Gravitino server can access them. Current Gravitino can support scripts in
 * the local file system, or on the web server (e.g., HTTP, HTTPS, FTP). Distributed file systems
 * like HDFS or S3 will be supported in the future.
 */
public class ShellJobTemplate extends JobTemplate {

  private final List<String> scripts;

  /**
   * Constructs a ShellJobTemplate with the specified builder.
   *
   * @param builder the builder containing the configuration for the shell job template
   */
  protected ShellJobTemplate(Builder builder) {
    super(builder);
    this.scripts = builder.scripts;
  }

  /**
   * Returns the list of scripts to be executed by the shell job template.
   *
   * @return the list of scripts
   */
  public List<String> scripts() {
    return scripts;
  }

  @Override
  public JobType jobType() {
    return JobType.SHELL;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof ShellJobTemplate)) return false;
    if (!super.equals(o)) return false;

    ShellJobTemplate that = (ShellJobTemplate) o;
    return Objects.equals(scripts, that.scripts);
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), scripts);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    if (!scripts.isEmpty()) {
      sb.append("  scripts=[\n");
      scripts.forEach(s -> sb.append("    ").append(s).append("\n"));
      sb.append("  ]\n");
    } else {
      sb.append("  scripts=[]\n");
    }

    return "\nShellJobTemplate{\n" + super.toString() + sb + "}\n";
  }

  /**
   * Creates a new builder for constructing instances of {@link ShellJobTemplate}.
   *
   * @return a new instance of {@link Builder}
   */
  public static Builder builder() {
    return new Builder();
  }

  /** Builder for creating instances of {@link ShellJobTemplate}. */
  public static class Builder extends BaseBuilder<Builder, ShellJobTemplate> {

    private List<String> scripts;

    private Builder() {}

    /**
     * Sets the scripts to be executed by the shell job template.
     *
     * @param scripts the list of scripts to be executed
     * @return the builder instance for method chaining
     */
    public Builder withScripts(List<String> scripts) {
      this.scripts = scripts;
      return this;
    }

    /**
     * Build the ShellJobTemplate instance.
     *
     * @return the constructed ShellJobTemplate
     */
    @Override
    public ShellJobTemplate build() {
      validate();
      return new ShellJobTemplate(this);
    }

    @Override
    protected Builder self() {
      return this;
    }

    @Override
    protected void validate() {
      super.validate();

      this.scripts = scripts != null ? ImmutableList.copyOf(scripts) : ImmutableList.of();
    }
  }
}

File: clients/client-python/gravitino/api/job/shell_job_template.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

from copy import deepcopy
from typing import List, Any
from .job_template import JobTemplate, JobType


class ShellJobTemplate(JobTemplate):
    """
    Represents a job template for executing shell scripts.
    """

    def __init__(self, builder: "ShellJobTemplate.Builder"):
        """
        Initializes a ShellJobTemplate instance.
        """
        super().__init__(builder)
        self._scripts = builder.scripts

    @property
    def scripts(self) -> List[str]:
        """
        Returns the list of shell scripts associated with this job template.

        Returns:
            List of shell scripts.
        """
        return self._scripts

    def job_type(self) -> JobType:
        """
        Returns the type of the job, which is SHELL for this template.

        Returns:
            JobType.SHELL: Indicates that this is a shell job template.
        """
        return JobType.SHELL

    @staticmethod
    def builder() -> "ShellJobTemplate.Builder":
        """
        Creates a new builder instance for constructing a ShellJobTemplate.

        Returns:
            ShellJobTemplate.Builder: A new builder instance.
        """
        return ShellJobTemplate.Builder()

    def __eq__(self, other: Any) -> bool:
        if not super().__eq__(other):
            return False
        return self._scripts == other.scripts

    def __hash__(self) -> int:
        return hash((super().__hash__(), tuple(self._scripts)))

    def __str__(self) -> str:
        res = "\nSellJobTemplate{\n"
        if self._scripts:
            res += (
                "  scripts=[\n"
                + "".join(f"    {s},\n" for s in self._scripts)
                + "  ],\n"
            )
        else:
            res += "  scripts=[],\n"

        return res + super().__str__() + "}\n"

    class Builder(
        JobTemplate.BaseBuilder["ShellJobTemplate.Builder", "ShellJobTemplate"]
    ):
        """
        Builder class for creating ShellJobTemplate instances.
        """

        def __init__(self):
            """
            Initializes the builder with default values.
            """
            super().__init__()
            self.scripts: List[str] = []

        def with_scripts(self, scripts: List[str]) -> "ShellJobTemplate.Builder":
            """
            Sets the shell scripts for this job template.

            Args:
                scripts: A list of shell scripts to be executed by this job template.

            Returns:
                ShellJobTemplate.Builder: The builder instance for method chaining.
            """
            self.scripts = deepcopy(scripts)
            return self

        def validate(self):
            """
            Validates the builder state before building the ShellJobTemplate.
            """
            super().validate()
            self.scripts = self.scripts or []

        def build(self) -> "ShellJobTemplate":
            """
            Builds and returns a ShellJobTemplate instance.

            Returns:
                ShellJobTemplate: The constructed shell job template.
            """
            self.validate()
            return ShellJobTemplate(self)
